### PR TITLE
Add n18401

### PR DESCRIPTION
**Pull request by jnewbery**

### CODE REVIEW

File: _posts/2020-04-01-#18401.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  amounts of all inputs (instead of just the amount of the input being signed)
+  and to the scriptPubKey of the output being spent. See [BIP
+  341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message)
+  for the full description of the transaction digest.
+
+- Because of these changes, more data needs to be stored in the
+  `PrecomputedTransactionData` object. This PR changes the way that the object
+  is constructed and initialized so that a [future
+  commit](https://github.com/bitcoin/bitcoin/pull/17977/commits/dd0622b8e2f8a324fdf73eb476f136ee556f3e68)
+  can add that data to the object.
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: suggest: a future commit in PR 17977

File: _posts/2020-04-01-#18401.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  we'll look at two more small refactor PRs that have also been pulled out of
+  that PR.
+
+- The same general considerations apply to these PRs: ordinarily, these
+  refactor changes to consensus code wouldn't meet the bar for review cost and
+  risk -vs- benefit.  However, splitting them off from the main schnorr/taproot
+  PR makes reviewing the logical changes to consensus in PR 17977 easier and
+  safer.
+
+### Initialize PrecomputedTransactionData in CheckInputScripts
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: suggest title these two PR headers the same as at the top by prefixing with "PR 18401: "

File: _posts/2020-04-01-#18401.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message)
+  for the full description of the transaction digest.
+
+- Because of these changes, more data needs to be stored in the
+  `PrecomputedTransactionData` object. This PR changes the way that the object
+  is constructed and initialized so that a [future
+  commit](https://github.com/bitcoin/bitcoin/pull/17977/commits/dd0622b8e2f8a324fdf73eb476f136ee556f3e68)
+  can add that data to the object.
+
+### Move single-sig checking EvalScript code to EvalChecksig
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: idem "PR 18422: Move..."

File: _posts/2020-04-01-#18401.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+- [Tapscript](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki)
+  is the scripting language used for scripts in a taproot tree (taproot has
+  versioning for scripts, so a future soft fork could introduce a different
+  scripting version or language).
+
+- Tapscript has almost the same semantics as Bitcoin script. One notable
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: s/script/Script/? (also as per https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md)

File: _posts/2020-04-01-#18401.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  difference is that `OP_CHECKSIG` and `OP_CHECKSIGVERIFY` both verify schnorr
+  signatures instead of ECDSA signatures.
+
+- This PR extracts most of the logic for `OP_CHECKSIG` and
+  `OP_CHECKSIGVEIFY` from the very large `EvalScript()` function.
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: seems to be a mix of DEIFY and VERIFY :)

File: _posts/2020-04-01-#18401.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+- A [future
+  commit](https://github.com/bitcoin/bitcoin/pull/17977/commits/af411a6cbec12375627aff374874a7f24b7c07e9)
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: suggest: a future commit in PR 17977

File: _posts/2020-04-01-#18401.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+1. Did you review the PRs? [Concept ACK, approach ACK, tested ACK, or
+   NACK?](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)
+   (Don't forget to put your PR review on GitHub.)
+
+2. A new `PrecomputedTransactionData` default constructor is added which
+   doesn't initialize the data members. Where do they get initialized instead?
+
+3. What new data members are added to `PrecomputedTransationData` in the
+   subsequent commit in the taproot PR?
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: maybe specify the PR 17977, or better yet, the commit (as there are 17 of them in 17977)

File: _posts/2020-04-01-#18401.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+4. Is the old `PrecomputedTransactionData` constructor still used anywhere?
+
+5. Why is the transaction digest algorithm changed for taproot? What problems
+   does the new algorith solve?
```
Comment (jonatack) [inline _posts/2020-04-01-#18401.md]: algorithm

File: _posts/2020-04-01-#18401.md (hunk 5 above)
Comment (jonatack) [reply _posts/2020-04-01-#18401.md]: ahhh, DEFI :)

File: _posts/2020-04-01-#18401.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  explanations.
+
+- To resolve the quadratic hashing problem, the way that the transaction digest
+  is calculated was changed for segwit inputs. See [BIP
+  143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki) for the
+  motivation and specification.
+
+- The segwit transaction digest is designed so parts of the digest are shared
+  between all the signatures in a transaction. This means that the total amount
+  of data to be hashed for the transaction is linear in the size of the
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply _posts/2020-04-01-#18401.md]: I _think_ this is right. We usually describe algorithms as 'linear/quadratic/exponential/etc in [variable]'

File: _posts/2020-04-01-#18401.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  transaction.
+
+- This shared data between all signatures of a transaction is stored in the
+  `PrecomputedTransactionData` object, which was added in [PR
+  8524](https://github.com/bitcoin/bitcoin/pull/8524).
+
+- Segwit v1 (taproot) makes some very minor changes to the transaction
+  digest algorithm. Specifically, the transaction digest commits to the
+  amounts of all inputs (instead of just the amount of the input being signed)
+  and to the scriptPubKey of the output being spent. See [BIP
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply _posts/2020-04-01-#18401.md]: see question 5

File: _posts/2020-04-01-#18401.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  of data to be hashed for the transaction is linear in the size of the
+  transaction.
+
+- This shared data between all signatures of a transaction is stored in the
+  `PrecomputedTransactionData` object, which was added in [PR
+  8524](https://github.com/bitcoin/bitcoin/pull/8524).
+
+- Segwit v1 (taproot) makes some very minor changes to the transaction
+  digest algorithm. Specifically, the transaction digest commits to the
+  amounts of all inputs (instead of just the amount of the input being signed)
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply _posts/2020-04-01-#18401.md]: see question 5

File: _posts/2020-04-01-#18401.md (hunk 7 above)
Comment (jnewbery) [reply _posts/2020-04-01-#18401.md]: Added the PR number, but left finding the exact commit as a (very easy - it's in the commit log summary) exercise for the reader.

**Comment by jnewbery** [comment]
Thanks @jonatack ! I've commented on a couple of your suggestions and taken all the rest.

**Pull request merged**